rfcs/convergence/as-prop.md

```diff
@@ -0,0 +1,197 @@
+      <CustomButton type="submit" />
+    </>
+  );
+}
+```
+
+**Retrospective note:** It was required to support React elements in `as` prop for [Semantic UI React](https://github.com/Semantic-Org/Semantic-UI-React) due elements composition:
+
+```tsx
+<>
+  {/* ✅ will render a proper markup expected by CSS styling */}
+  {/* <div class="ui sidebar menu"/> */}
+  <Sidebar as={Menu} />
+
+  {/* ❌ styles will be different in this case */}
+  {/* <div class="ui sidebar"><div class="ui menu" /></div> */}
+  <Sidebar>
+    <Menu />
+  </Sidebar>
+</>
+```
+
+On Fluent side we don't have such restrictions from styling system.
+
+## Event handlers problem
+
+_It's a real scenario from Teams._
+
+Without this change we know that in `as` we can get a memoized component that forces us memoize event handlers:
+
+```tsx
+function Button() {
+  // 💣 Oops, this will break memoized input, should be wrapped with React.useCallback()
+  const handleClick = () => {};
+
+  return <Element onClick={handleClick} />;
+}
+
+<Button as={MemoEl} />;
+```
+
+This can be extremely tricky for us and customers with hooks approach as they may not be aware of these requirements:
+
+```tsx
+function useButton(state) {
+  state.onClick = React.useCallback(/* some code */);
+}
+function useCustomButton(state) {
+  // 💣 this is not memoized anymore
+  state.onClick = () => {
+    state.onClick();
+    /* some custom interactions */
+  };
+}
+```
+
+## Accessibility problem
+
+Another problem is related to accessibility handling as we need to override attributes or key handling based on a passed element, for example:
+
+```tsx
+<>
+  {/* renders <button /> */}
+  <Button />
+  {/* renders <div role="button" /> */}
+  <Button as="div" />
+</>
+```
+
+```ts
+// A pseudocode to show possible logic
+if (props.as !== 'button') {
+  state.role = 'button';
+}
+```
+
+However, if user will pass a React component we don't know HTML tag before rendering:
+
+```tsx
+// 🤔 What renders "SomeComponent"?!
+<Button as={SomeComponent} />
+```
+
+To solve this we can use tag detection in effects (like [Reakit does](https://github.com/reakit/reakit/blob/a211d94da9f3b683182568a56479b91afb1b85ae/packages/reakit/src/Button/Button.ts#L34-L39)), but it will cause a double render for `<Button as="div" />`:
+
+- render a component
+- resolve effects: check tagName and apply roles if needed, trigger an update
+- render a component again
+
+## Detailed Design or Proposal
+
+Restrict `as` to HTML tags:
+
+```diff
+-  as?: React.ElementType;
++  as?: keyof JSX.IntrinsicElements;
+```
+
+With this change we will allow to pass only HTML tags to `as` prop:
+
+```tsx
+<>
+  {/* ✅ renders <span /> */}
+  <Text as="span" />
+  {/* ❌ throws a compiler error */}
+  <Button as={Link} />
+</>
+```
+
+Scenarios with React elements are not so frequent, in this case is proposed to compose components instead:
+
+```tsx
+function LinkButton() {
+  const { state } = useButton();
+
+  return <Link {...state.root} />;
```

Comment:
won't this also throw a compiler error if button state can't be coerced to link state ?

Reply:
As we have there `any` currently - nope, but it's a good question... Probably, it should something like:

```tsx
function LinkButton() {
  const { state } = useButton();
  const nativeProps = getNativeProps(state.root, anchorProperties);

  return <Link {...nativeProps} />;
}
```

It will be more detailed in upcoming RFC related slots, as ideally it should something like this:

```tsx
function LinkButton(props: ButtonProps & LinkProps) {
  const { state, render } = useButton({ components: { root: Link } });

  return render(state);
}
```

Reply:
The `LinkButton` example for me would make a lot more sense, why can't that be the example ?

Reply:
And what is wrong with `<Link as="button" />

Reply:
It's not an actual API, but agree, pushed a change 👍